src/encoded/upgrade/genetic_modification.py

```diff
@@ -185,4 +185,78 @@ def genetic_modification_5_6(value, system):
+
+    sources = {
+        'addgene': r'^\d{5,6}$',
+        'bacpac': r'^([A-Z]{2,3}\d{2,3}|[A-Z]{3})-\d{1,4}[A-Z]\d{1,2}$',
+        'dharmacon': r'^[DL]-\d{6}-\d{2}(-\d{2,4})?$',
+        'hugo-bellen': r'^MI\d{5}$',
+        'human-orfeome': r'^([A-Z]{2})?\d{1,9}$',
+        'plasmid-repository': r'^HsCD\d{8}$',
+        'sigma': r'^[A-Z]{3}\d{3}$',
+        'source-bioscience': r'^[A-Z]{3}\d{3,4}[a-z][A-Z]\d{2}(_[A-Z]\d{2})?$',
+        'thermo-fisher': r'^[a-zA-Z]{1,3}\d{5,6}$',
+        'trc': r'^TRCN\d{10}$',
+        'brenton-graveley': r'^BGC#\d{7}$'
+    }
+    reagents = value.get('reagents')
+    if reagents:
+        new_reagents = []
+        unmatched_reagents = ''
+        for reagent in reagents:
+            identifier = reagent['identifier']
+            source_id = reagent['source']
+            source_obj = conn.get_by_uuid(source_id)
+            source_name = source_obj.properties['name']
+            new_reagent = copy.deepcopy(reagent)
+            matching = []
+            source_from_reagent = None
+            for source in sources:
+                if source in source_name:
+                    source_from_reagent = source_name
+                    break
+            # Preferentially add prefix to reagent that matches the reagent source
+            if source_from_reagent is not None:
+                regex = sources[source_from_reagent]
+                if re.match(regex, identifier) is not None:
+                    matching.append((source_name, source_from_reagent, identifier))
+            else:
+                for source, regex in sources.items():
+                    if re.match(regex, identifier) is not None:
+                        matching.append((source_name, source, identifier))
+            if not matching:
+                unmatched_reagents += '{} '.format(json.dumps(reagent))
+            else:
+                if len(matching) >= 2:
+                    # Bring the identifier with matching source to the reagent to the top
+                    matching.sort(key=lambda match: int(match[1] in match[0]), reverse=True)
+                    # matching = [i for i in matching if i[1] in i[0]]
+                _, source, identifier = matching[0]
+                new_identifier = ('{}:{}'.format(source, identifier))
+                new_reagent['identifier'] = new_identifier
+                new_reagents.append(new_reagent)
+        # Dump the unmatched identifier reagents to the notes
+        if unmatched_reagents:
+            date = time.strftime('%m/%d/%Y')
+            message = 'On {}, the following reagents were removed due to invalid identifer: '.format(date)
+            if 'notes' in value:
+                new_notes = value['notes'] + ' ' + message + unmatched_reagents
+            else:
+                new_notes = message + unmatched_reagents
+            new_notes = new_notes.strip()
+            value['notes'] = new_notes
+        del value['reagents']
```

Comment:
This is kinda weird hanging out here all alone.  Do we delete the reagents if there are not any new_reagents?  Maybe we should delete them at the top, I think there is a way to pop from a dict instead of 'reagents = value.get('reagents')'  and then delete.

Reply:
Yeah we still want to delete them if there aren''t any new ones. .pop is a good idea, I'll add that in.